### DISCUSSION TITLE
Nestjs mikro-orm respository usage in tests

### DISCUSSION THREAD

**Discussion by maciossek** [mikro-orm/mikro-orm#6195] [Q&A]
Hi there,

I was refactoring my code when I encountered that I am not fully aware of how to inject entity repositories in nestjs test modules.

```typescript
import { EntityRepository } from "@mikro-orm/core";
import { getMikroORMToken, getRepositoryToken } from "@mikro-orm/nestjs";
import { INestApplication } from "@nestjs/common";
import { Test } from "@nestjs/testing";
import request from "supertest";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { DatabaseContext } from "../../../../database/databaseContext.enum";
import { ApplicationFactory } from "../../../../database/postgresql-default/seeder/factories";
import { Application } from "../../applications.entity";
import { AppRole } from "../app-role.entity";
import { AppRolesController } from "../app-roles.controller";
import { AppRolesService } from "../app-roles.service";

describe("AppRolesController GET /:uuid", () => {
  let app: INestApplication;
  let applicationFactory: ApplicationFactory;
  let applicationRepository: EntityRepository<Application>;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [global.mikroOrmPostgreSqlModule],
      controllers: [AppRolesController],
      providers: [
        AppRolesService,
      ],
    }).compile();

    app = moduleRef.createNestApplication();
    await app.init();

    const orm = app.get(getMikroORMToken(DatabaseContext.default));
    applicationRepository = orm.em.getRepository(Application);
    applicationFactory = new ApplicationFactory(orm.em);
  });

  afterEach(async () => {
    await applicationRepository.nativeDelete({});
  });


  afterAll(async () => {
    await app.close();
  });
});
```

throws `Error: Nest can't resolve dependencies of the AppRolesService (?). Please make sure that the argument "AppRoleRepository_default" at index [0] is available in the RootTestModule context.`

so far I have been using the entityManager in my `AppRolesService` and that worked perfectly. 

I tried with 
```typescript
const moduleRef = await Test.createTestingModule({
  imports: [global.mikroOrmPostgreSqlModule],
  controllers: [AppRolesController],
  providers: [
    AppRolesService,
    {
      provide: getRepositoryToken(AppRole),
      useClass: EntityRepository,
    },
  ],
}).compile();
```

But with no success.

My workaround is

```typescript
@Injectable()
export class AppRolesService {
  // constructor(
  //   @InjectRepository(AppRole, DatabaseContext.default)
  //   private readonly appRoleRepository: EntityRepository<AppRole>,
  // ) {}

  private appRoleRepository: EntityRepository<AppRole>;

  constructor(
    @InjectEntityManager(DatabaseContext.default)
    private readonly em: EntityManager,
  ) {
    this.appRoleRepository = em.getRepository(AppRole);
  }
}
```

Any help much appreciated <3

**Comment by B4nan** [comment]
I can't help much with this myself, so let's hope someone who uses nest will also chip in, but at least a few notes from my end.

>     {
>      provide: getRepositoryToken(AppRole),
>      useClass: EntityRepository,
>    },

This is surely wrong, you can't just expect nest DI to create a repository for you, only by providing `useClass: EntityRepository` - that is a generic class, nest wont understand how to instantiate it, and even if it could understand that it requires the EM in its constructor, it also needs the entity name.

Normally, you need `forFeature` calls, those register the repositories. And those services are registered as factories, depending on the EM service, using `em.getRepository()` for the value. So if something, you might do this:

https://github.com/mikro-orm/nestjs/blob/master/src/mikro-orm.providers.ts#L125-L127

Or just add the `forFeature` call to your test module, I would say that is the correct solution unless you want to mock the repository.
